File: src/web/src/types/post.ts
// Post types
export type PostType = 'Cycle' | 'Inventory' | 'Feedback';

export type FeedbackCategory =
  | 'General'
  | 'Cycles'
  | 'Inventory'
  | 'Tumblers'
  | 'Gallery'
  | 'Materials'
  | 'Specimens'
  | 'FAQ'
  | 'Settings';

export interface PostDto {
  postId: string;
  userId: string;
  cycleId: string | null;
  inventoryId: string | null;
  postType: PostType;
  feedbackCategory: FeedbackCategory | null;
  title: string;
  description: string | null;
  status: string;
  publishedDate: string;
  voteCount: number;
  commentCount: number;
  author: PostAuthorDto;
  cycle: CyclePreviewDto | null;
  inventory: InventoryPreviewDto | null;
  photos: PostPhotoDto[];
}

export interface PostListDto {
  postId: string;
  postType: PostType;
  feedbackCategory: FeedbackCategory | null;
  title: string;
  description: string | null;
  publishedDate: string;
  voteCount: number;
  commentCount: number;
  author: PostAuthorDto;
  coverPhotoUrl: string | null;
  coverPhotoThumbnailUrl: string | null;
  coverPhotoBlurHash: string | null;
  photoCount: number;
  // Inventory-specific fields for gallery card display
  sourceType: string | null;
  specimenNames: string[];
  sizeCategories: string[];
}

export interface PostAuthorDto {
  userId: string;
  username: string;
  displayName: string | null;
  avatarUrl: string | null;
}

export interface PostPhotoDto {
  postId: string;
  photoId: string;
  url: string;
  sortOrder: number;
  isCover: boolean;
  thumbnailUrl: string | null;
  mediumUrl: string | null;
  largeUrl: string | null;
  blurHash: string | null;
  width: number | null;
  height: number | null;
}

export interface CyclePreviewDto {
  cycleId: string;
  name: string;
  status: string;
  startDate: string;
  endDate: string | null;
  difficultyRating: number | null;
  finalQuality: number | null;
  stageCount: number;
  // Extended fields for gallery display
  elapsedDays: number;
  totalRuntimeHours: number;
  photoCount: number;
  tumblerName: string | null;
  barrelName: string | null;
  specimenNames: string[];
}

export interface InventoryPreviewDto {
  inventoryId: string;
  name: string;
  sourceType: string;
  sourceName: string | null;
  sourceLocation: string | null;
  acquiredDate: string;
  condition: string;
  specimenNames: string[];
  sizeCategories: string[];
  photoCount: number;
}

export interface CreatePostRequest {
  cycleId?: string;
  inventoryId?: string;
  feedbackCategory?: FeedbackCategory;
  title: string;
  description?: string;
  photoIds?: string[];
  coverPhotoId?: string;
}

export interface UpdatePostRequest {
  title: string;
  description?: string;
}

// Vote types
export interface VoteDto {
  voteId: string;
  postId: string;
  userId: string;
  dateCreated: string;
}

export interface VoteCountDto {
  postId: string;
  count: number;
  userHasVoted: boolean;
}

// Comment types
export interface CommentDto {
  commentId: string;
  postId: string;
  parentCommentId: string | null;
  content: string;
  isEdited: boolean;
  editedDate: string | null;
  dateCreated: string;
  author: CommentAuthorDto;
  replies: CommentDto[];
}

export interface CommentAuthorDto {
  userId: string;
  username: string;
  displayName: string | null;
  avatarUrl: string | null;
}

export interface CreateCommentRequest {
  content: string;
  parentCommentId?: string;
}

export interface UpdateCommentRequest {
  content: string;
}

export interface ReportCommentRequest {
  reason: string;
  details?: string;
}

File: src/api/MyUglyRocks.Core/Entities/Post.cs
namespace MyUglyRocks.Core.Entities;

public enum PostStatus
{
    Published = 0,
    Hidden = 1,
    Removed = 2
}

public enum PostType
{
    Cycle = 0,
    Inventory = 1,
    Feedback = 2
}

public enum FeedbackCategory
{
    General = 0,
    Cycles = 1,
    Inventory = 2,
    Tumblers = 3,
    Gallery = 4,
    Materials = 5,
    Specimens = 6,
    FAQ = 7,
    Settings = 8
}

public class Post : SoftDeletableEntity
{
    public Guid PostId { get; set; }
    public Guid UserId { get; set; }

    // PostType determines whether this is a Cycle post, Inventory post, or Feedback post
    public PostType PostType { get; set; } = PostType.Cycle;

    // For Cycle and Inventory posts: Either CycleId or InventoryId must be set
    // For Feedback posts: Both should be null
    public Guid? CycleId { get; set; }
    public Guid? InventoryId { get; set; }

    // For Feedback posts: Category is required
    // For Cycle/Inventory posts: Category is null
    public FeedbackCategory? FeedbackCategory { get; set; }

    public required string Title { get; set; }
    public string? Description { get; set; }
    public PostStatus Status { get; set; } = PostStatus.Published;
    public DateTime PublishedDate { get; set; }
    public int VoteCount { get; set; }
    public int CommentCount { get; set; }

    // Navigation properties
    public virtual User User { get; set; } = null!;
    public virtual Cycle? Cycle { get; set; }
    public virtual Inventory? Inventory { get; set; }
    public virtual ICollection<PostPhoto> PostPhotos { get; set; } = [];
    public virtual ICollection<Vote> Votes { get; set; } = [];
    public virtual ICollection<Comment> Comments { get; set; } = [];
}
